Documentation: Manual memory handling explained

PallHaraldsson opened the pull request:
I'm not up to speed on @doc and not sure if I should have added here:

https://github.com/JuliaLang/julia/blob/master/base/libc.jl

PallHaraldsson committed:
Documentation: Manual memory handling explained [ci skip]

I'm not up to speed on @doc and not sure if I should have added here:

https://github.com/JuliaLang/julia/blob/master/base/libc.jl

yuyichao commented on doc/stdlib/libc.rst:
```diff
@@ -16,6 +16,8 @@
    Call ``calloc`` from the C standard library.
 
+   See warning in the documentation for ``free`` regarding only using this on memory originally obtained from ``malloc``\ .
+
```
Hmm.... ? I don't think this applies

yuyichao commented on doc/stdlib/libc.rst:
```diff
@@ -8,7 +8,7 @@
    .. Docstring generated from Julia source
 
-   Call ``malloc`` from the C standard library.
+   Call ``malloc`` from the C standard library. Note, memory allocated this way, using the S Standard Library, will result in a memory leak (not handled by Julia's garbage collection), when memory is handled in this manual way. This use, is for Julia's internal standard library use (or those who know what they are doing). See also the embedding section of the manual before proceding, if your inte ...
```
"S Standard Library"?

Also, this is not just for "internal standard library".

PallHaraldsson replied:
Sorry, typo for C (will fix), thanks for confirming. I put in parens "or those who know what they are doing".

PallHaraldsson committed:
Update libc.rst

PallHaraldsson commented:
Thanks for the help.. I'm just trying to help and learn in the way. :)

PallHaraldsson committed:
Maybe even better text..

yuyichao commented on doc/stdlib/libc.rst:
```diff
@@ -8,7 +8,7 @@
    .. Docstring generated from Julia source
 
-   Call ``malloc`` from the C standard library.
+   Call ``malloc`` from the C standard library. Note, this is not the standard way to allocate memory in Julia (there is no need to allocate or deallocate manually). This indirect way of "manual memory allocation", using the C Standard Library, will result in a memory leak if the memory is not freed later explicitly, e.g. it will not be handled by Julia's garbage collection mechanism. This use, i ...
```
It might be nice to emphasize the need of manually freeing the memory although I personally feel like `C malloc` already cover that all.

I'm not sure what is the opposite of `not the standard way`. Also not sure what do you mean by `indirect way of`. I still don't think you should mention `for ... internal ... use` because there's nothing internal about it. This is also not related to embedding or the julia GC (which is described in the embedding section and I guess that you mean by `embedding section of the manual`). This is completely independent of the GC and you shouldn't need to understand the Julia GC in order to use `malloc`. It doesn't give you "more control" either.

To summary, this is allocating c memory and not GC memory so the difference is the type of memory, not how standard/direct/controllable they are. You use this if the c library requires it and that's pretty much it. I don't really think the text you've added explains that better than just ``call `malloc` in libc``.

yuyichao commented:
Also, see https://github.com/JuliaLang/julia/blob/6f3ccdda604be8c7254c63306e55ffd6074898b7/CONTRIBUTING.md#improving-documentation for how you should update the function doc.

PallHaraldsson committed:
Fixed (I think) based on feedback

"embedding section and I guess that's you mean by embedding section of the manual" yes, except I was probably thinking of http://docs.julialang.org/en/release-0.4/manual/calling-c-and-fortran-code/

Anyway, I dropped the language until I'm sure it's better to include. Whta I had in mind with "more control" (also dropped), is if you could abuse Julia by using (only) memory management, I guess you would need to disabe the GC..

PallHaraldsson commented:
@yuyichao Should we just close this?

I was just reviewing my old PRs and I can't locate where libc.rst has gone.
